File: Raven.Client.Java/lib/src/main/java/raven/abstractions/connection/OAuthHelper.java
package raven.abstractions.connection;

import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

import org.apache.commons.codec.binary.Base64;
import org.apache.commons.lang.ArrayUtils;
import org.apache.commons.lang.StringUtils;

import raven.client.utils.encryptors.Encryptor;
import raven.client.utils.encryptors.IAsymmetricalEncryptor;
import raven.client.utils.encryptors.IHashEncryptor;
import raven.client.utils.encryptors.ISymmetricalEncryptor;

public class OAuthHelper {
  public static class Keys {
    public static String EncryptedData = "data";
    public static String APIKeyName = "api key name";
    public static String Challenge = "challenge";
    public static String Response = "response";

    public static String RSAExponent = "exponent";
    public static String RSAModulus = "modulus";

    public static String ChallengeTimestamp = "pepper";
    public static String ChallengeSalt = "salt";
    public static int ChallengeSaltLength = 64;

    public static String ResponseFormat = "%s;%s";
    public static String WWWAuthenticateHeaderKey = "Raven ";
  }

  public static String hash(String data) {
    IHashEncryptor sha1 = Encryptor.getCurrent().createHash();
    return bytesToString(sha1.compute20(data.getBytes()));
  }

  public static String encryptAsymmetric(byte[] exponent, byte[] modulus, String data) {
    try {
      byte[] bytes = data.getBytes();

      ISymmetricalEncryptor aesKeyGen = Encryptor.getCurrent().createSymmetrical(256);
      aesKeyGen.generateKey();
      aesKeyGen.generateIV();

      byte[] encryptedKeyAndIv = addEncryptedKeyAndIv(exponent, modulus, aesKeyGen.getKey(), aesKeyGen.getIV());

      Cipher encryptor = aesKeyGen.createEncryptor();
      byte[] encryptedBytes = encryptor.doFinal(bytes);

      return bytesToString(ArrayUtils.addAll(encryptedKeyAndIv, encryptedBytes));
    } catch (BadPaddingException | IllegalBlockSizeException | NoSuchAlgorithmException | InvalidKeyException | NoSuchPaddingException | InvalidAlgorithmParameterException e) {
      throw new RuntimeException(e);
    }
  }

  private static byte[] addEncryptedKeyAndIv(byte[] exponent, byte[] modulus, byte[] key, byte[] iv) {
    try {
      IAsymmetricalEncryptor rsa = Encryptor.getCurrent().createAsymmetrical(exponent, modulus);
      return rsa.encrypt(ArrayUtils.addAll(key, iv));
    } catch (NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException | NoSuchPaddingException | IllegalBlockSizeException | BadPaddingException e) {
      throw new RuntimeException(e);
    }
  }

  public static Map<String, String> parseDictionary(String data) {
    String[] tokens = data.split(",");
    Map<String, String> result = new HashMap<>();
    for (String token: tokens) {
      String[] items = StringUtils.splitPreserveAllTokens(token,'=');
      if (items.length > 2) {
        String key = items[0];
        items = Arrays.copyOfRange(items, 1, items.length);
        result.put(key.trim(), StringUtils.join(items, "=").trim());
      } else {
        result.put(items[0].trim(), items[1].trim());
      }
    }
    return result;
  }

  public static String dictionaryToString(Map<String, String> data) {
    List<String> items = new ArrayList<>();
    for (Map.Entry<String, String> entry: data.entrySet()) {
      items.add(entry.getKey() + "=" + entry.getValue());
    }
    return StringUtils.join(items, ",");
  }

  public static byte[] parseBytes(String data) {
    if (data == null) {
      return null;
    }
    return Base64.decodeBase64(data);
  }

  public static String bytesToString(byte[] data) {
    if (data == null) {
      return null;
    }
    return Base64.encodeBase64String(data);
  }


  public static void main(String[] args) throws Exception {
    //TODO: delete me
//    OAuthHelper.encryptAsymmetric(exponent, modulus, data)

    String encodeBase64 = Base64.encodeBase64String(new byte[] { 1,2,3});

    String modulus = "oLG/fa+JA7scV7rj290GM651EzyrAHDG6vyzPwqxAwNgfcTzG4i9RuQ41q+So4Qg9NX1oQHq4ieYx9NjTAeM4qzIQR5QJnj2i4D6FLWMxzUdkliiBC2GDJ/5uxFYK/TPJMvg7+yhayZcXeuiRq8KtK1GBLiAeb1Q3kIEvxn1M6eGgK134Zbxtr8skuViKqxMkJw3OJrqI/IuMnW9aEHEEyA9kjVXokgx1C0u0ypaLxpKCj0mu/KAWc7r5x4M2qjSd0G6TVea4suTt5UM5/Pq/DhmPu2jGHAwxMJ5rfqXRUBtYirgJ/KqIZlZMWuHSbBMUg0aSOmikPycv3uUIbgSbw";
    String exp = "AQAB";

    byte[] decodeModulus = Base64.decodeBase64(modulus);
    byte[] decodeExp = Base64.decodeBase64(exp);

    IAsymmetricalEncryptor encryptor = Encryptor.getCurrent().createAsymmetrical(decodeExp, decodeModulus);
    byte[] encrypt = encryptor.encrypt(new byte[] { 1,2,3});


    String asymmetric = OAuthHelper.encryptAsymmetric(decodeExp, decodeModulus, "XXXX");
    System.out.println(asymmetric);
  }
}

File: src/Raven.NewClient/Connection/OAuthHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Raven.NewClient.Abstractions.Extensions;

namespace Raven.NewClient.Abstractions.Connection
{
    using Raven.NewClient.Abstractions.Util.Encryptors;

    public static class OAuthHelper
    {
        public static class Keys
        {
            public const string EncryptedData = "data";
            public const string APIKeyName = "api key name";
            public const string Challenge = "challenge";
            public const string Response = "response";

            public const string RSAExponent = "exponent";
            public const string RSAModulus = "modulus";

            public const string ChallengeTimestamp = "pepper";
            public const string ChallengeSalt = "salt";
            public const int ChallengeSaltLength = 64;

            public const string ResponseFormat = "{0};{1}";
            public const string WWWAuthenticateHeaderKey = "Raven ";
        }

        [ThreadStatic]
        private static IHashEncryptor sha1;

        /**** Cryptography *****/

        public static string Hash(string data)
        {
            var bytes = Encoding.UTF8.GetBytes(data);

            if (sha1 == null)
                sha1 = Encryptor.Current.CreateHash();
            var hash = sha1.Compute20(bytes);
            return BytesToString(hash);
        }

        public static string EncryptAsymmetric(byte[] exponent, byte[] modulus, string data)
        {
            var bytes = Encoding.UTF8.GetBytes(data);
            var results = new List<byte>();

            using (var aesKeyGen = Encryptor.Current.CreateSymmetrical(keySize: 256))
            {
                aesKeyGen.GenerateKey();
                aesKeyGen.GenerateIV();

                results.AddRange(AddEncryptedKeyAndIv(exponent, modulus, aesKeyGen.Key, aesKeyGen.IV));

                using (var encryptor = aesKeyGen.CreateEncryptor())
                {
                    var encryptedBytes = encryptor.TransformEntireBlock(bytes);
                    results.AddRange(encryptedBytes);
                }
            }
            return BytesToString(results.ToArray());
        }

        private static byte[] AddEncryptedKeyAndIv(byte[] exponent, byte[] modulus, byte[] key, byte[] iv)
        {
            using (var rsa = Encryptor.Current.CreateAsymmetrical(exponent, modulus))
            {
                return rsa.Encrypt(key.Concat(iv).ToArray());
            }
        }

        /**** On the wire *****/

        public static Dictionary<string, string> ParseDictionary(string data)
        {
            return data.Split(',')
                .Select(item =>
                {
                    var items = item.Split(new[] { '=' }, StringSplitOptions.None);
                    if (items.Length > 2)
                    {
                        return new[] { items[0], string.Join("=", items.Skip(1)) };
                    }
                    return items;
                })
                .ToDictionary(
                    item => (item.First()).Trim(),
                    item => (item.Skip(1).FirstOrDefault() ?? "").Trim()
                );
        }

        public static string DictionaryToString(Dictionary<string, string> data)
        {
            return string.Join(",", data.Select(item => item.Key + "=" + item.Value));
        }

        public static byte[] ParseBytes(string data)
        {
            if (data == null)
                return null;
            return Convert.FromBase64String(data);
        }

        public static string BytesToString(byte[] data)
        {
            if (data == null)
                return null;
            return Convert.ToBase64String(data);
        }
    }
}
